From 283f1bceb87883ba6c965750e9880ce3d67af089 Mon Sep 17 00:00:00 2001
From: Stefan Bodewig <bodewig@apache.org>
Date: Tue, 4 Jul 2017 11:07:16 +0200
Subject: [PATCH] COMPRESS-416 disable powermock test on Java9

---
 .../compressors/z/ZCompressorInputStreamTest.java      | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/src/test/java/org/apache/commons/compress/compressors/z/ZCompressorInputStreamTest.java b/src/test/java/org/apache/commons/compress/compressors/z/ZCompressorInputStreamTest.java
index 7b9ed54d5af..8f62c9ba642 100644
--- a/src/test/java/org/apache/commons/compress/compressors/z/ZCompressorInputStreamTest.java
+++ b/src/test/java/org/apache/commons/compress/compressors/z/ZCompressorInputStreamTest.java
@@ -39,6 +39,14 @@ public class ZCompressorInputStreamTest {
 
     @Test(expected = IOException.class)
     public void testFailsToCreateZCompressorInputStreamAndThrowsIOException() throws IOException {
+        boolean java9 = false;
+        try {
+            Class.forName("java.lang.module.ModuleDescriptor");
+            java9 = true;
+        } catch (Exception ex) {
+            // not Java9
+        }
+        org.junit.Assume.assumeFalse("can't use PowerMock with Java9", java9);
 
         Enumeration<SequenceInputStream> enumeration = (Enumeration<SequenceInputStream>) mock(Enumeration.class);
         SequenceInputStream sequenceInputStream = new SequenceInputStream(enumeration);
@@ -51,4 +59,4 @@ public void testFailsToCreateZCompressorInputStreamAndThrowsIOException() throws
     }
 
 
-}
\ No newline at end of file
+}
