From 3fd2492fcce073f0c36e4d7e23e34881557e6e5e Mon Sep 17 00:00:00 2001
From: Yuming Wang <wgyumg@gmail.com>
Date: Mon, 4 Jun 2018 23:52:28 +0800
Subject: [PATCH] PARQUET-1317: Fix ParquetMetadataConverter throw NPE (#489)

---
 .../parquet/format/converter/ParquetMetadataConverter.java    | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java b/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
index 40c0b841fc..2baad15cd3 100644
--- a/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
+++ b/parquet-hadoop/src/main/java/org/apache/parquet/format/converter/ParquetMetadataConverter.java
@@ -1170,9 +1170,9 @@ private void buildChildren(Types.GroupBuilder builder,
       }
       if (schemaElement.isSetConverted_type()) {
         LogicalTypeAnnotation originalType = getOriginalType(schemaElement.converted_type, schemaElement);
-        LogicalTypeAnnotation newLogicalType = getOriginalType(schemaElement.logicalType);
+        LogicalTypeAnnotation newLogicalType = schemaElement.isSetLogicalType() ? getOriginalType(schemaElement.logicalType) : null;
         if (!originalType.equals(newLogicalType)) {
-          childBuilder.as(getOriginalType(schemaElement.converted_type, schemaElement));
+          childBuilder.as(originalType);
         }
       }
       if (schemaElement.isSetField_id()) {
