From abc2d23900ff5adeef648879af6bc4d2f79c261e Mon Sep 17 00:00:00 2001
From: Stefan Bodewig <bodewig@apache.org>
Date: Sat, 11 Jun 2016 17:33:15 +0200
Subject: [PATCH] COMPRESS-357 synchronize the important part of finish again

---
 .../bzip2/BZip2CompressorOutputStream.java    | 27 ++++++++++---------
 1 file changed, 15 insertions(+), 12 deletions(-)

diff --git a/src/main/java/org/apache/commons/compress/compressors/bzip2/BZip2CompressorOutputStream.java b/src/main/java/org/apache/commons/compress/compressors/bzip2/BZip2CompressorOutputStream.java
index 3bdee5aa366..034fbfc20b3 100644
--- a/src/main/java/org/apache/commons/compress/compressors/bzip2/BZip2CompressorOutputStream.java
+++ b/src/main/java/org/apache/commons/compress/compressors/bzip2/BZip2CompressorOutputStream.java
@@ -321,6 +321,7 @@ private static void hbMakeCodeLengths(final byte[] len, final int[] freq,
     private Data data;
     private BlockSort blockSorter;
 
+    private final Object outLock = new Object();
     private OutputStream out;
 
     /**
@@ -476,21 +477,23 @@ protected void finalize() throws Throwable {
     }
 
 
-    public synchronized void finish() throws IOException {
-        if (out != null) {
-            try {
-                if (this.runLength > 0) {
-                    writeRun();
+    public void finish() throws IOException {
+        synchronized(outLock) {
+            if (out != null) {
+                try {
+                    if (this.runLength > 0) {
+                        writeRun();
+                    }
+                    this.currentChar = -1;
+                    endBlock();
+                    endCompression();
+                } finally {
+                    this.out = null;
                 }
-                this.currentChar = -1;
-                endBlock();
-                endCompression();
-            } finally {
-                this.out = null;
-                this.data = null;
-                this.blockSorter = null;
             }
         }
+        this.blockSorter = null;
+        this.data = null;
     }
 
     @Override
